packages/vite/src/node/server/moduleGraph.ts

```diff
@@ -237,10 +236,16 @@ export class ModuleGraph {
     url = removeImportQuery(removeTimestampQuery(url))
     const resolved = await this.resolveId(url, !!ssr)
     const resolvedId = resolved?.id || url
```

Comment:
`resolveUrl` is called with both rollup-space URL and browser-space URL.
Is this function intended to be called by both? Would this need to be resolved to a same url? (currently it doesn't)

You can check this by adding this line here and running `playground/resolve`.
```js
if (url.includes('@virtual-file')) console.trace(url, resolved?.id, resolved?.id.replace('\0', '!0!'))
```

Reply:
I think it isn't clear, it is a similar context to `ssrLoadModule` as explained in https://github.com/vitejs/vite/pull/9124#issuecomment-1186777686. In Vite 4 we could review if we should for `resolveUrl` to always be in a given space, but since `ssrLoadModule` already accepts both, maybe we should also unwrap the id and replace to `\0` as a guard at the start of it? If not we should review where is it being called with a URL out in browser space and pre-unwrap it.
Looks like this ensure has receiving browser-space URLs since before the release of Vite 2. See https://github.com/vitejs/vite/commit/8d8e2cc9ea0e1cc8f620c0a9d143294c1efb37b0#diff-f2f744fef86a2c562dd5142240912f7a2d28404fac536740a2424daf628aa609R208

I don't know what is the best here during v3. The module graph is user-facing, so if we start accepting both then it may be hard to properly decode/encode. 

Maybe we could still try to keep this function as is, and pre-unwrap in the places we are missing? We could already create the `toRollupURL` helper and use it for these places?

For reference, in the module graph we have:
- `urlToModuleMap` that maps a Rollup-space URL to a Module
- `idToModuleMap` that maps a resolved id to a Module

Since we are calling `resolveUrl` with both internally, we could review if what we really want in the graph is Browser-space URL or Rollup-space URLs, but since in both transformRequest and in ssrLoadModule we are using Rollup-space URLs when calling it I think it the above is fine and we should expect to see `\0` and unwrapped ids (no `/@fs/`, etc) when exploring the module graph by URL. cc @antfu, just in case you have seen this in your plugins.

Reply:
Umm, it seems it requires much more work. I feel I need to read though all the code if I'm going to tackle this.
I'll leave this one as-is in this PR.

Reply:
Yes, and maybe it is better, as it is unrelated to the fix this PR is targeting.
We can continue to work on this on future smaller PRs, and do a full review for 3.1.
Thanks a lot @sapphi-red!